manager/shard_strategy.go

```diff
@@ -0,0 +1,226 @@
+package manager
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/flyteorg/flytepropeller/manager/config"
+	"github.com/flyteorg/flytepropeller/pkg/apis/flyteworkflow/v1alpha1"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+// Defines necessary functionality for a sharding strategy
+type ShardStrategy interface {
+	// Returns the total number of pods for the sharding strategy
+	GetPodCount() int
+	// Generates a unique hash code to identify shard strategy updates
+	HashCode() (uint32, error)
+	// Updates the PodSpec for the specified index to include label selectors
+	UpdatePodSpec(pod *v1.PodSpec, containerName string, podIndex int) error
+}
+
+// Creates and validates a new ShardStrategy defined by the configuration
```

Comment:
```suggestion
// NewShardStrategy creates and validates a new ShardStrategy defined by the configuration
```

same here. Also repeat this for all docs. Lets make propeller even better documented